wasm-backend: Code compiled with fPIC is not compatible with regular executables

See the corresponding upstream llvm issue: https://bugs.llvm.org/show_bug.cgi?id=42656

I've tested LLVM 032e1fdcd47ef98936c88aa30571c6a713179a46 with a build of https://zlib.net/zlib-1.2.11.tar.gz . It throws: 

```
wasm-ld: /llvm-10.0/llvm-project.git/lld/wasm/Symbols.cpp:115: void lld::wasm::Symbol::setGOTIndex(uint32_t): Assertion `gotIndex == INVALID_INDEX' failed.
Stack dump:
0.      Program arguments: /llvm-10.0/bin/wasm-ld -o libz.so.1.2.11 --allow-undefined --import-memory --import-table --lto-O0 adler32.lo crc32.lo deflate.lo infback.lo inffast.lo inflate.lo inftrees.lo trees.lo zutil.lo compress.lo uncompr.lo gzclose.lo gzlib.lo gzread.lo gzwrite.lo -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr --export __wasm_call_ctors --export __data_end --export main --export malloc --export free --export setThrew --export __errno_location -z stack-size=5242880 --initial-memory=16777216 --no-entry --max-memory=16777216 --global-base=1024 --relocatable
 #0 0x00007fd2cb76c1f9 llvm::sys::PrintStackTrace(llvm::raw_ostream&) /llvm-10.0/llvm-project.git/llvm/lib/Support/Unix/Signals.inc:532:0
 #1 0x00007fd2cb76c28c PrintStackTraceSignalHandler(void*) /llvm-10.0/llvm-project.git/llvm/lib/Support/Unix/Signals.inc:593:0
 #2 0x00007fd2cb76a15b llvm::sys::RunSignalHandlers() /llvm-10.0/llvm-project.git/llvm/lib/Support/Signals.cpp:68:0
 #3 0x00007fd2cb76bbb0 SignalHandler(int) /llvm-10.0/llvm-project.git/llvm/lib/Support/Unix/Signals.inc:384:0
 #4 0x00007fd2cadf2890 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x12890)
 #5 0x00007fd2c9eaee97 raise /build/glibc-OTsEL5/glibc-2.27/signal/../sysdeps/unix/sysv/linux/raise.c:51:0
 #6 0x00007fd2c9eb0801 abort /build/glibc-OTsEL5/glibc-2.27/stdlib/abort.c:81:0
 #7 0x00007fd2c9ea039a __assert_fail_base /build/glibc-OTsEL5/glibc-2.27/assert/assert.c:89:0
 #8 0x00007fd2c9ea0412 (/lib/x86_64-linux-gnu/libc.so.6+0x30412)
 #9 0x00007fd2cbb82930 lld::wasm::Symbol::setGOTIndex(unsigned int) /llvm-10.0/llvm-project.git/lld/wasm/Symbols.cpp:116:0
#10 0x00007fd2cbba5b0b lld::wasm::GlobalSection::assignIndexes() /llvm-10.0/llvm-project.git/lld/wasm/SyntheticSections.cpp:243:0
#11 0x00007fd2cbb88fdd (anonymous namespace)::Writer::assignIndexes() /llvm-10.0/llvm-project.git/lld/wasm/Writer.cpp:624:0
#12 0x00007fd2cbb8ad23 (anonymous namespace)::Writer::run() /llvm-10.0/llvm-project.git/lld/wasm/Writer.cpp:899:0
#13 0x00007fd2cbb8b7f0 lld::wasm::writeResult() /llvm-10.0/llvm-project.git/lld/wasm/Writer.cpp:983:0
#14 0x00007fd2cbb54594 (anonymous namespace)::LinkerDriver::link(llvm::ArrayRef<char const*>) /llvm-10.0/llvm-project.git/lld/wasm/Driver.cpp:791:0
#15 0x00007fd2cbb4fad7 lld::wasm::link(llvm::ArrayRef<char const*>, bool, llvm::raw_ostream&) /llvm-10.0/llvm-project.git/lld/wasm/Driver.cpp:95:0
#16 0x00007fd2cb6fa9e5 main /llvm-10.0/llvm-project.git/lld/tools/lld/lld.cpp:159:0
#17 0x00007fd2c9e91b97 __libc_start_main /build/glibc-OTsEL5/glibc-2.27/csu/../csu/libc-start.c:344:0
#18 0x00007fd2cb6f9b2a _start (/llvm-10.0/bin/wasm-ld+0x2f9b2a)
shared:ERROR: '/llvm-10.0/bin/wasm-ld -o libz.so.1.2.11 --allow-undefined --import-memory --import-table --lto-O0 adler32.lo crc32.lo deflate.lo infback.lo inffast.lo inflate.lo inftrees.lo trees.lo zutil.lo compress.lo uncompr.lo gzclose.lo gzlib.lo gzread.lo gzwrite.lo -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr --export __wasm_call_ctors --export __data_end --export main --export malloc --export free --export setThrew --export __errno_location -z stack-size=5242880 --initial-memory=16777216 --no-entry --max-memory=16777216 --global-base=1024 --relocatable' failed (-6)
Makefile:282: recipe for target 'libz.so.1.2.11' failed```

--- a/emscripten.py
+++ b/emscripten.py
@@ -2225,10 +2225,11 @@ def emscript_wasm_backend(infile, outfile, memfile, compiler_engine,
 
   asm_consts, asm_const_funcs = create_asm_consts_wasm(forwarded_json, metadata)
   em_js_funcs = create_em_js(forwarded_json, metadata)
+  asm_const_pairs = ['%s: %s' % (key, value) for key, value in asm_consts]
+  asm_const_map = 'var ASM_CONSTS = {\n  ' + ',  \n '.join(asm_const_pairs) + '\n};\n'
   pre = pre.replace(
     '// === Body ===',
-    ('// === Body ===\n\nvar ASM_CONSTS = [' +
-     ',\n '.join(asm_consts) + '];\n' +
+    ('// === Body ===\n\n' + asm_const_map +
      asstr('\n'.join(asm_const_funcs)) +
      '\n'.join(em_js_funcs) + '\n'))
   pre = apply_table(pre)
@@ -2342,7 +2343,7 @@ def finalize_wasm(temp_files, infile, outfile, memfile, DEBUG):
 
 
 def create_asm_consts_wasm(forwarded_json, metadata):
-  asm_consts = [0] * len(metadata['asmConsts'])
+  asm_consts = {}
   all_sigs = []
   for k, v in metadata['asmConsts'].items():
     const, sigs, call_types = v
@@ -2417,6 +2418,8 @@ function %s(code, sigPtr, argbuf) {%s
   var args = readAsmConstArgs(sigPtr, argbuf);
   return ASM_CONSTS[code].apply(null, args);
 }''' % (const_name, preamble))
+  asm_consts = [(key, value) for key, value in asm_consts.items()]
+  asm_consts.sort()
   return asm_consts, asm_const_funcs
 
 

--- a/tests/test_other.py
+++ b/tests/test_other.py
@@ -8684,8 +8684,8 @@ int main() {
 var ASM_CONSTS = [function() { var x = !<->5.; }];
                                         ^
 ''', '''
-var ASM_CONSTS = [function() {var x = !<->5.;}];
-                                       ^
+  0: function() {var x = !<->5.;}
+                          ^
 '''), stderr)
 
   def test_EM_ASM_ES6(self):

